infrastructures/infrastructure-factory/src/main/java/org/eclipse/che/api/factory/server/scm/kubernetes/KubernetesGitCredentialManager.java

```diff
@@ -0,0 +1,167 @@
+
+import com.google.common.collect.ImmutableMap;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.api.model.ObjectMetaBuilder;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.SecretBuilder;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.factory.server.scm.GitCredentialManager;
+import org.eclipse.che.api.factory.server.scm.PersonalAccessToken;
+import org.eclipse.che.api.factory.server.scm.exception.ScmConfigurationPersistenceException;
+import org.eclipse.che.api.factory.server.scm.exception.UnsatisfiedScmPreconditionException;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.commons.lang.NameGenerator;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.api.shared.KubernetesNamespaceMeta;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory;
+
+/**
+ * Creates or updates Git credentials secret in user-s namespace to allow Git operations on private
+ * repositories.
+ */
+@Singleton
+public class KubernetesGitCredentialManager implements GitCredentialManager {
+  public static final String NAME_PATTERN = "%s-git-credentials-secret-";
+  public static final String ANNOTATION_SCM_URL = "che.eclipse.org/scm-url";
+  public static final String ANNOTATION_SCM_USERNAME = "che.eclipse.org/scm-username";
+  public static final String ANNOTATION_CHE_USERID = "che.eclipse.org/che-userid";
+
+  private static final Map<String, String> LABELS =
+      ImmutableMap.of(
+          "app.kubernetes.io/part-of", "che.eclipse.org",
+          "app.kubernetes.io/component", "workspace-secret");
+
+  private static final Map<String, String> ANNOTATIONS =
+      ImmutableMap.of(
+          ANNOTATION_AUTOMOUNT,
+          "true",
+          ANNOTATION_MOUNT_PATH,
+          "/home/theia/.git-credentials",
+          ANNOTATION_MOUNT_AS,
+          "file",
+          ANNOTATION_GIT_CREDENTIALS,
+          "true");
+
+  private final KubernetesNamespaceFactory namespaceFactory;
+  private final KubernetesClientFactory clientFactory;
+
+  @Inject
+  public KubernetesGitCredentialManager(
+      KubernetesNamespaceFactory namespaceFactory, KubernetesClientFactory clientFactory) {
+    this.namespaceFactory = namespaceFactory;
+    this.clientFactory = clientFactory;
+  }
+
+  @Override
+  public void createOrReplace(PersonalAccessToken personalAccessToken)
+      throws UnsatisfiedScmPreconditionException, ScmConfigurationPersistenceException {
+    try {
+      String namespace = getFirstNamespace();
+      Optional<Secret> existing =
```

Comment:
Could we explain with a comment why we need to fetch the secret if we "createOrReplace" it later on? Why not just "createOrReplace" the secret constructed anew from the personal access token? My guess would be to not destroy the custom labels/annos put on it by the user. Is that right? Or are there other reasons?

Reply:
There is two reasons:
1) We can't have two different git secrets in namespace now, and 
2) secrets have random part in their names
so we try to reuse existing one by matching hostname/username if possible, and update it. Otherwise, create new one.

Reply:
could that be a comment in the code?

Reply:
added comment